test/lit/passes/optimize-instructions-sign_ext.wast

```diff
@@ -0,0 +1,185 @@
+;; NOTE: Assertions have been generated by update_lit_checks.py and should not be edited.
+;; RUN: wasm-opt %s -all --optimize-instructions -S -o - | filecheck %s
+
+(module
+  ;; CHECK:      (memory $0 16 17)
+  (memory $0 16 17)
+
+  ;; CHECK:      (func $i32-direct (type $1) (param $x i32)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (i32.extend8_s
+  ;; CHECK-NEXT:    (local.get $x)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $i32-direct (param $x i32)
+    ;; We do not need to sign-extend twice, and can emit just one extend8.
+    (drop
+      (i32.extend8_s
+        (i32.shr_s
+          (i32.shl
+            (local.get $x)
+            (i32.const 24)
+          )
+          (i32.const 24)
+        )
+      )
+    )
+  )
+
+  ;; CHECK:      (func $i32-local (type $0)
+  ;; CHECK-NEXT:  (local $temp i32)
+  ;; CHECK-NEXT:  (local.set $temp
+  ;; CHECK-NEXT:   (i32.load8_s
+  ;; CHECK-NEXT:    (i32.const 22)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (local.get $temp)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $i32-local
+    ;; The local is sign-extended, so the i32.extend can be removed.
+    (local $temp i32)
+    (local.set $temp
+      (i32.load8_s
+        (i32.const 22)
+      )
+    )
+    (drop
+      (i32.extend8_s
+        (local.get $temp)
+      )
+    )
+  )
+
+  ;; CHECK:      (func $i32-local-i16 (type $0)
+  ;; CHECK-NEXT:  (local $temp i32)
+  ;; CHECK-NEXT:  (local.set $temp
+  ;; CHECK-NEXT:   (i32.load16_s
+  ;; CHECK-NEXT:    (i32.const 22)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (local.get $temp)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $i32-local-i16
+    ;; As above with i16.
+    (local $temp i32)
+    (local.set $temp
+      (i32.load16_s
+        (i32.const 22)
+      )
+    )
+    (drop
+      (i32.extend16_s
+        (local.get $temp)
+      )
+    )
+  )
+
+  ;; CHECK:      (func $i32-local-i16-bad (type $0)
+  ;; CHECK-NEXT:  (local $temp i32)
+  ;; CHECK-NEXT:  (local.set $temp
+  ;; CHECK-NEXT:   (i32.load8_s
+  ;; CHECK-NEXT:    (i32.const 22)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (i32.extend16_s
+  ;; CHECK-NEXT:    (local.get $temp)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $i32-local-i16-bad
+    ;; As above with in i8/i16 mismatch. We do not optimize.
```

Comment:
This should still be optimizable, right? If the value is already 8-bit sign extended, then further 16-bit sign extension won't change it. It's the other order of operations that would be a problem.

Reply:
Makes sense, it seems worth optimizing those. Done with a little refactoring.